netmaster/master/policy.go

```diff
@@ -0,0 +1,436 @@
+/***
+Copyright 2014 Cisco Systems Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package master
+
+import (
+	"errors"
+	"strconv"
+
+	"github.com/contiv/objmodel/contivModel"
+	"github.com/contiv/ofnet"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+// Create the netmaster
+var ofnetMaster *ofnet.OfnetMaster
+
+// InitPolicyMgr initializes the policy manager
+func InitPolicyMgr() error {
+	ofnetMaster = ofnet.NewOfnetMaster(ofnet.OFNET_MASTER_PORT)
+	if ofnetMaster == nil {
+		log.Fatalf("Error creating ofnet master")
+	}
+
+	return nil
+}
+
+// getEndpointGroupID returns endpoint group Id for a service
+// It autocreates the endpoint group if it doesnt exist
+func getEndpointGroupID(serviceName, networkName string) (int, error) {
+	// FIXME: Get tenant name
+	tenantName := "default"
+
+	// If service name is not specified, we are done
+	if serviceName == "" {
+		return 0, nil
+	}
+
+	// form the key based on network and service name.
+	epgName := serviceName + "." + networkName
+	epgKey := tenantName + ":" + epgName
+
+	// See if the epg exists
+	epg := contivModel.FindEndpointGroup(epgKey)
+	if epg == nil {
+		endpointGroup := contivModel.EndpointGroup{
+			Key:         epgKey,
+			TenantName:  tenantName,
+			NetworkName: networkName,
+			GroupName:   epgName,
+		}
+
+		// Create endpoint group
+		err := contivModel.CreateEndpointGroup(&endpointGroup)
+		if err != nil {
+			log.Errorf("Error creating endpoint group: %+v, Err: %v", endpointGroup, err)
+			return 0, err
+		}
+
+		// find again
+		epg = contivModel.FindEndpointGroup(epgKey)
+		if epg == nil {
+			log.Errorf("Error creating endpoint group %s.", epgName)
+			return 0, errors.New("EPG not created")
+		}
+	}
+
+	// return endpoint group id
+	return epg.EndpointGroupID, nil
+}
+
+// PolicyAttach attaches a policy to an endpoint and adds associated rules to policyDB
+func PolicyAttach(epg *contivModel.EndpointGroup, policy *contivModel.Policy) error {
+	epgpKey := epg.Key + ":" + policy.Key
+
+	// See if it already exists
+	gp := FindEpgPolicy(epgpKey)
+	if gp != nil {
+		log.Errorf("EPG policy %s already exists", epgpKey)
+		return errors.New("Epg policy exists")
```

Comment:
I think @mapuri means that we can use the core.Error struct in place of `error` (`fmt.Errorf` or `errors.New`) in our new errors. It implements the `error` interface so I think we'll be ok.

It is a convention in this project, I think it's best we stick with it.